File: medium/matrix_islands/matrix_islands.py
def matrix_islands(grid):
    """
    grid: Matrix of ints
    Returns the number of "islands" in grid
    """
    if not grid:
        return 0
    r = len(grid)
    c = len(grid[0])
    visited = [[False for _ in range(c)] for _ in range(r)]
    
    
    def visit(i, j):
        if i < 0 or j < 0 or i >= r or j >= c:
            return
        if visited[i][j] or grid[i][j] == 0:
            return
        visited[i][j] = True
        
        visit(i + 1, j)
        visit(i, j + 1)
        visit(i - 1, j)
        visit(i, j - 1)
    
    result = 0
    for i in range(r):
        for j in range(c):
            if not visited[i][j] and grid[i][j] == 1:
                result += 1
                visit(i, j)
                
    return result

File: medium/matrix_islands/matrix_islands.js
function visit(visited, mat, i, j) {
  if (i < 0 || j < 0 || i >= mat.length || j >= mat[0].length) {
    return;
  }

  if (visited[i][j] || !mat[i][j]) {
    return;
  }

  visited[i][j] = 1;
  visit(visited, mat, i + 1, j);
  visit(visited, mat, i, j + 1);
  visit(visited, mat, i - 1, j);
  visit(visited, mat, i, j - 1);
}

/**
 *
 * @param {number[][]} mat
 * @returns {number}
 */
function matrix_islands(mat) {
  let islandCount = 0;

  const visited = Array(mat.length)
    .fill()
    .map(() => Array(mat[0].length).fill(0));

  for (let i = 0; i < mat.length; i++) {
    for (let j = 0; j < mat[i].length; j++) {
      if (!visited[i][j] && mat[i][j]) {
        islandCount++;
        visit(visited, mat, i, j);
      }
    }
  }
  return islandCount;
}

module.exports = matrix_islands;
